### ISSUE TITLE
dart2wasm fails on Random.secure().nextInt()

### ISSUE THREAD

**Issue by konsultaner** [dart-lang/sdk#55031]
At the moment it is not possible to use the `Random.secure().nextInt()` method in conjunction with dart2wasm.

```dart
import 'dart:math';

import 'package:test/test.dart';

void main() {
  group('Dart2WASM', () {
    test('Random secure issue', () async {
      expect(Random.secure().nextInt(256), 2);
    });
  });
}
```
this test running with `-p chrome -c dart2wasm`

will output:

```
unparsed  unparsed      at _SecureRandom._getBytes (http://localhost:38411/iGXEHWjpHrlKzYPYwWHrFGDKxWMbPDMS/test/issue_test.dart.browser_test.dart.wasm:wasm-function[1593]:0x56673)
unparsed  unparsed      at new _SecureRandom (constructor body) (http://localhost:38411/iGXEHWjpHrlKzYPYwWHrFGDKxWMbPDMS/test/issue_test.dart.browser_test.dart.wasm:wasm-function[1615]:0x570bd)
unparsed  unparsed      at _SecureRandom (http://localhost:38411/iGXEHWjpHrlKzYPYwWHrFGDKxWMbPDMS/test/issue_test.dart.browser_test.dart.wasm:wasm-function[1613]:0x570b0)
unparsed  unparsed      at Random._secureRandom (http://localhost:38411/iGXEHWjpHrlKzYPYwWHrFGDKxWMbPDMS/test/issue_test.dart.browser_test.dart.wasm:wasm-function[1612]:0x57091)
unparsed  unparsed      at new Random.secure (http://localhost:38411/iGXEHWjpHrlKzYPYwWHrFGDKxWMbPDMS/test/issue_test.dart.browser_test.dart.wasm:wasm-function[822]:0x433b1)
unparsed  unparsed      at main closure at file://##############.dart:7:61 inner (http://localhost:38411/iGXEHWjpHrlKzYPYwWHrFGDKxWMbPDMS/test/issue_test.dart.browser_test.dart.wasm:wasm-function[819]:0x43335)
unparsed  unparsed      at main closure at file://##############.dart:7:61 (http://localhost:38411/iGXEHWjpHrlKzYPYwWHrFGDKxWMbPDMS/test/issue_test.dart.browser_test.dart.wasm:wasm-function[812]:0x4321c)

NoSuchMethodError: method not found: '_getBytes'
Receiver: null
Arguments: []

```

**Comment by konsultaner** [comment]
I'm not a pro for writing wasm interop code but it seems like

https://github.com/dart-lang/sdk/blob/4642c73347c3d9e04b0d5a4078bd4ba40cfd56d3/sdk/lib/_internal/wasm/lib/math_patch.dart#L246

is missing a

````dart
@pragma("wasm:import", "get_random_bytes")
```

I found this docu about the wasm random:

https://github.com/WebAssembly/wasi-random?tab=readme-ov-file#api-walk-through

Also there are a lot of todos in the math class. Is it fully supported yet?

@mkustermann You seem to be working on wasm. I would contribute, but it's quite hard to get into the code. Is there a guide how the wasm binding works with dart?

**Comment by mkustermann** [comment]
Thank you for the bug report, @konsultaner !

This is a bug: It hasn't been implemented yet (the only `external` method in that patch that has no implementation) - it was probably missed.

We have a test for this: https://github.com/dart-lang/sdk/blob/main/tests/lib/math/random_secure_test.dart which is failing on dart2wasm atm.

> Also there are a lot of todos in the math class. Is it fully supported yet?

Those TODOs aren't critical to address (in fact they come from the VM backend - which the wasm version was based on - I believe)

Generally most things should be working in dart2wasm. There's a few things that aren't quite there yet (e.g. edge cases in async, https://github.com/dart-lang/sdk/issues/55025), which we'll work on in the coming weeks/months.

Though please notice that dart2wasm is still experimental

> @pragma("wasm:import", "get_random_bytes")

Our main target for dart2wasm is the browser atm, which doesn't have WASI support. I think maybe something like [this](https://www.w3schools.com/jsref/met_crypto_getrandomvalues.asp) will do the trick.

/cc @osa1 Could you implement `Random.secure()`?

**Comment by osa1** [comment]
Implemented in https://dart-review.googlesource.com/c/sdk/+/356040.

**Comment by konsultaner** [comment]
@osa1 Should I close this issue as you implemented it?

**Comment by osa1** [comment]
@konsultaner it's not ready yet. The issue will be automatically closed when it's merged.

**Comment by konsultaner** [comment]
>Implemented in https://dart-review.googlesource.com/c/sdk/+/356040

@osa1 which version of dart will have it included?

**Comment by osa1** [comment]
@konsultaner the commit 12859761b9d46b14e7f33c7433ddd3e991709a54 was tagged as 3.4.0-dev so it will be included in 3.4.0.
